examples/fluids/src/petsc_ops.c

```diff
@@ -323,7 +321,7 @@ PetscErrorCode CreateSolveOperatorsFromMatCeed(KSP ksp, Mat mat_ceed, PetscBool
   }
 
   if (assemble_amat) {
-    PetscCall(MatConvert(mat_ceed, mat_ceed_inner_type, MAT_INITIAL_MATRIX, Amat));
+    PetscCall(MatCeedSetupInnerMat(mat_ceed, Amat));
```

Comment:
We definitely don't want to `MatConvert` here for sure. We don't want to lose the `mat_ceed`.

This not what `MatCeedSetupInnerMat` is for per the description so I'd be really careful making a function public that was designed to only be working on internal data. The language around "InnerMat" would be very confusing.

How about

- Convert `MatCeedSetupInnerMat` to `MatCeedCreateMatCOO`
- Rename option `ceed_inner_mat_type` to `ceed_coo_mat_type`
- Split the COO setup out of `MatCeedAssembleCOO` into `MatCeedSetPreallocationCOO`
- Call `MatCeedSetPreallocationCOO` in both `MatCeedAssembleCOO` and `MatCeedCreateMatCOO`

Thoughts?

Reply:
> We definitely don't want to `MatConvert` here for sure. We don't want to lose the `mat_ceed`.

Well, `MatConvert` doesn't lose `mat_ceed`; it generates an entirely new matrix. We just can't use that matrix for COO assembly. Or at least that's my interpretation of the errors that came from trying to do that.

> This not what `MatCeedSetupInnerMat` is for per the description so I'd be really careful making a function public that was designed to only be working on internal data. The language around "InnerMat" would be very confusing.

Yeah, I initially tried to make a new function, but couldn't figure out what the structure should be (or the naming) and just went with publicizing `MatCeedSetupInnerMat` as the simplest proof-of-concept.

> 
> How about
>  * Convert `MatCeedSetupInnerMat` to `MatCeedCreateMatCOO`

:+1: 

>  * Rename option `ceed_inner_mat_type` to `ceed_coo_mat_type`

:+1: 

>  * Split the COO setup out of `MatCeedAssembleCOO` into `MatCeedSetPreallocationCOO`
>  * Call `MatCeedSetPreallocationCOO` in both `MatCeedAssembleCOO` and `MatCeedCreateMatCOO`

I think I prefer keeping assembly out of `MatCeedCreateMatCOO`. Based on other errors I was encountering, the assembly of our Jacobian matrix can't happen until the IFunction is run (and `jac_data` is populated) and the matrix creation needs to happen before TSSolve. I think also having `MatCeedCreateMatCOO` sounds more analogous to `DMCreateMat`, which also just creates an "empty" `Mat` object.

Reply:
Right, I'm proposing that assembly doesn't happen during `MatCeedCreateMatCOO`, only setting the COO pattern

Reply:
Ope, misread. I guess I'm not sure why we'd want to have the COO preallocation in both `MatCeedCreateMatCOO` and `MatCeedAssembleCOO`? Leaving it only in `MatCeedAssembleCOO` is the simpler change; we can always migrate to COO-setup in both later.

Reply:
My thinking was that you could View the Mat with the sparsity pattern already set out of the box. I can show you what I'm thinking in a quick Ratel MR - its the same complexity either way.

Reply:
https://gitlab.com/micromorph/ratel/-/merge_requests/831

Reply:
https://github.com/CEED/libCEED/pull/1574